### PR TITLE
Refactor Partial Hydration

### PR DESCRIPTION
**Pull request by sebmarkbage**
This is a pretty invasive refactor of the SuspenseComponent and DehydratedSuspenseComponent.

The primary purpose of this refactor is to avoid the hacky "upgrade" and "downgrade" by mutating the tag.

In the new model, a boundary is always represented by the same SuspenseComponent fiber. Inside it I store a "DehydratedFragment". The inner fiber represents the dehydrated nodes in the tree. This can be used to delete the whole thing or insert before it.

I also switched legacy mode to always client-render the boundary content since we can't hydrate partially in legacy mode. It also warns.

Otherwise, this (hopefully) shouldn't have an semantic differences.

### CODE REVIEW

File: packages/react-reconciler/src/ReactFiberBeginWork.js [hunk 1]
```diff
@@ -1646,26 +1643,121 @@ function updateSuspenseComponent(
...
     // ensure the state of the primary children is preserved.
-    const prevState = current.memoizedState;
-    const prevDidTimeout = prevState !== null;
-    if (prevDidTimeout) {
+    const prevState: null | SuspenseState = current.memoizedState;
+    if (prevState !== null) {
+      if (enableSuspenseServerRenderer) {
+        const dehydrated = prevState.dehydrated;
+        if (dehydrated !== null) {
+          if (!didSuspend) {
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactFiberBeginWork.js]: More comments

File: scripts/error-codes/codes.json [hunk 2]
```diff
@@ -338,5 +338,7 @@
   "337": "An invalid event responder was provided to host component",
   "338": "ReactDOMServer does not yet support the fundamental API.",
   "339": "An invalid value was used as an event responder. Expect one or many event responders created via React.unstable_createResponer().",
-  "340": "An invalid value was used as an event listener. Expect one or many event listeners created via React.unstable_useResponer()."
+  "340": "An invalid value was used as an event listener. Expect one or many event listeners created via React.unstable_useResponer().",
```
Comment (acdlite) [inline scripts/error-codes/codes.json]: @trueadm Noticed a typo here. Should be `unstable_useResponder()`

Comment (trueadm) [reply scripts/error-codes/codes.json]: I’m on PTO. cc @necolas

Comment (acdlite) [reply scripts/error-codes/codes.json]: Oh yeah, sorry. I mean I can do it too :D

File: packages/react-reconciler/src/ReactFiberBeginWork.js [hunk 3]
```diff
@@ -1511,21 +1511,23 @@ function updateSuspenseComponent(
...
     // rendering the fallback children.
-    nextState = SUSPENDED_MARKER;
     nextDidTimeout = true;
     workInProgress.effectTag &= ~DidCapture;
   } else {
     // Attempting the main content
-    if (current === null || current.memoizedState !== null) {
+    if (
+      current === null ||
+      (current.memoizedState: null | SuspenseState) !== null
```
Comment (bvaughn) [inline packages/react-reconciler/src/ReactFiberBeginWork.js]: Why is this necessary? ^

File: scripts/error-codes/codes.json (hunk 2 above)
Comment (acdlite) [reply scripts/error-codes/codes.json]: https://github.com/facebook/react/pull/16373

File: packages/react-reconciler/src/ReactFiberBeginWork.js (hunk 3 above)
Comment (sebmarkbage) [reply packages/react-reconciler/src/ReactFiberBeginWork.js]: It's so that it becomes greppable.

Once we add types for fields like memoizedState it will also be used to ensure that those types are correct.

Comment (sebmarkbage) [reply packages/react-reconciler/src/ReactFiberBeginWork.js]: The type SuspenseState should probably be nullable by itself since it has meaning and would be a variant if we had those.
